### DISCUSSION TITLE
How do I automatically generate row numbers in DataGrid?

### DISCUSSION THREAD

**Discussion by howzew** [AvaloniaUI/Avalonia#17671] [Q&A]
I have the following code:

IMultiValueConverter：

```
    internal class RowToIndexConverter : IMultiValueConverter
    {
        public object Convert(IList<object?> values, Type targetType, object parameter, CultureInfo culture)
        {
            if (values == null || values.Count == 0 || values[0] == null)
            {
                return null;
            }

            if (values[0] is DataGridRow row)
            {
                if (row.DataContext?.GetType().FullName == "Avalonia.Controls.ItemsControl+None")
                {
                    return null;
                }

                return (row.GetIndex() + 1).ToString();
            }

            return null;
        }


        public object[] ConvertBack(object value, Type[] targetTypes, object parameter, CultureInfo culture)
        {
            throw new NotSupportedException();
        }
    }
```

Views:

```
	<DataGridTemplateColumn Header="No." Width="*" >
		<DataGridTemplateColumn.CellTemplate>
			<DataTemplate>
				<TextBlock HorizontalAlignment="Center">
					<TextBlock.Text>
						<MultiBinding Converter="{StaticResource RowToIndexConverter}">
							<Binding RelativeSource="{RelativeSource AncestorType={x:Type DataGridRow}}"/>
							<Binding RelativeSource="{RelativeSource AncestorType={x:Type DataGrid}}" Path="$self"/>
						</MultiBinding>
					</TextBlock.Text>
				</TextBlock>
			</DataTemplate>
		</DataGridTemplateColumn.CellTemplate>
	</DataGridTemplateColumn>
```

The above code can automatically add the row number after adding a new row in DataGrid, but when a row is deleted, the row number will not be automatically updated。

For example, there are 3 rows now, the row number is 1, 2, 3, I want to delete the second row, the row number becomes 1, 2, but it actually becomes 1, 3, how to optimize the code?

**Comment by howzew** [comment]
I've tried everything but it doesn't work

**Comment by maxkatz6** [comment]
Usually, it would be done with LoadingRow event and setting Row.Header.
Something like in this example:
https://stackoverflow.com/questions/4661998/simple-way-to-display-row-numbers-on-wpf-datagrid

**Reply by howzew** [reply]
I used this approach initially, but after using avalonia, I'm not sure if this is the most logical approach with avalonia MVVM.

**Reply by timunie** [reply]
In case of MVVM, you can also add a property RowNumber to ItemViewModel and update it in the ViewModel whenever collection changes. Just a for loop over each item should be enough.

**Reply by howzew** [reply]
Yes, this is a good idea, I'm just not sure if the loop will be inefficient if there are many elements in the collection

**Comment by rabbitism** [comment]
In fact my PR is already merged, you can now show Index in header easily with a style. Check the sample in ControlCatalog. 
https://github.com/AvaloniaUI/Avalonia/blob/18fcfcc16348401e74fd74abba11223e6f7b4fb0/samples/ControlCatalog/Pages/DataGridPage.xaml#L48-L52

**Reply by howzew** [reply]
Thank you very much, but the sequence number will start from 0, how can I start from 1?

**Reply by maxkatz6** [reply]
Add a converter
